src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Indentation/AbstractIndentation.cs

```diff
@@ -2,32 +2,36 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading;
 using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.Text;
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Indentation
 {
-    internal abstract partial class AbstractIndentationService<TSyntaxRoot> : IIndentationService
+    internal abstract partial class AbstractIndentation<TSyntaxRoot>
         where TSyntaxRoot : SyntaxNode, ICompilationUnitSyntax
     {
+        protected abstract ISyntaxFacts SyntaxFacts { get; }
+        protected abstract IHeaderFacts HeaderFacts { get; }
+        protected abstract ISyntaxFormatting SyntaxFormatting { get; }
+
         protected abstract AbstractFormattingRule GetSpecializedIndentationFormattingRule(FormattingOptions2.IndentStyle indentStyle);
-        protected abstract ISmartTokenFormatter CreateSmartTokenFormatter(Document document, TSyntaxRoot root, TextLine lineToBeIndented, IndentationOptions options);
 
-        private IEnumerable<AbstractFormattingRule> GetFormattingRules(Document document, int position, FormattingOptions2.IndentStyle indentStyle)
-        {
-            var workspace = document.Project.Solution.Workspace;
-            var formattingRuleFactory = workspace.Services.GetRequiredService<IHostDependentFormattingRuleFactoryService>();
-            var baseIndentationRule = formattingRuleFactory.CreateRule(document, position);
+        /// <summary>
+        /// Returns <see langword="true"/> if the language specific <see
+        /// cref="ISmartTokenFormatter"/> should be deferred to figure out indentation.  If so, it
+        /// will be asked to <see cref="ISmartTokenFormatter.FormatTokenAsync"/> the resultant
+        /// <paramref name="token"/> provided by this method.
+        /// </summary>
+        protected abstract bool ShouldUseTokenIndenter(Indenter indenter, out SyntaxToken token);
+        protected abstract ISmartTokenFormatter CreateSmartTokenFormatter(
+            TSyntaxRoot root, TextLine lineToBeIndented, IndentationOptions options, AbstractFormattingRule baseFormattingRule);
 
-            var formattingRules = new[] { baseIndentationRule, this.GetSpecializedIndentationFormattingRule(indentStyle) }.Concat(Formatter.GetDefaultFormattingRules(document));
-            return formattingRules;
-        }
+        protected abstract IndentationResult? GetDesiredIndentationWorker(
+            Indenter indenter, SyntaxToken? token, SyntaxTrivia? trivia);
 
+#if false
```

Comment:
WIP?

Reply:
Ah, is this because you are retaining `GetIndentation` on the `AbstractIndentationService`, which has just been moved? It seems fine to delete this block completely, unless you are trying to strip it down further to non-Workspace entities to move it down here.

Reply:
yup.  this is delete later.